pkg/serviceaccount/claims_test.go

```diff
@@ -0,0 +1,216 @@
+package serviceaccount
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/apis/core"
+
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+func init() {
+	now = func() time.Time {
+		// 1514764800
+		return time.Date(2018, time.January, 1, 0, 0, 0, 0, time.UTC)
+	}
+}
+
+func Test(t *testing.T) {
+	cs := []struct {
+		// input
+		sa  core.ServiceAccount
+		pod *core.Pod
+		sec *core.Secret
+		exp int64
+		aud []string
+		// desired
+		sc *jwt.Claims
+		pc *privateClaims
+	}{
+		{
+			// pod and secret
+			sa: core.ServiceAccount{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "a",
+					Name:      "b",
+					UID:       "c",
+				},
+			},
+			pod: &core.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "d",
+					Name:      "e",
+					UID:       "f",
+				},
+			},
+			sec: &core.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "g",
+					Name:      "h",
+					UID:       "k",
+				},
+			},
+			// really fast
+			exp: 0,
+			// nil audience
+			aud: nil,
+
+			sc: &jwt.Claims{
+				Subject:   "system:serviceaccount:a:b",
+				IssuedAt:  jwt.NumericDate(1514764800),
+				NotBefore: jwt.NumericDate(1514764800),
+				Expiry:    jwt.NumericDate(1514764800),
+			},
+			pc: &privateClaims{
+				Kubernetes: kubernetes{
+					Namespace: "a",
+					Svcacct: ref{
+						Name: "b",
+						UID:  "c",
+					},
+					Pod: &ref{
+						Name: "e",
+						UID:  "f",
+					},
+				},
+			},
+		},
+		{
+			// pod
+			sa: core.ServiceAccount{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "a",
+					Name:      "b",
+					UID:       "c",
+				},
+			},
+			pod: &core.Pod{
+				ObjectMeta: metav1.ObjectMeta{
```

Comment:
We're creating `metav1.ObjectMeta` 8 times here. Consider introducing a helper method for it:. `createObjectMeta("d", "e", "f")` will be shorter and easy to read.